src/NHibernate/Multi/QueryBatch.cs

```diff
@@ -120,53 +122,51 @@ private void Init()
 			}
 		}
 
-		private void CombineQueries(IResultSetsCommand resultSetsCommand)
+		private void ExecuteBatched()
 		{
-			foreach (var multiSource in _queries)
-			foreach (var cmd in multiSource.GetCommands())
+			var statsEnabled = Session.Factory.Statistics.IsStatisticsEnabled;
+			Stopwatch stopWatch = null;
+			if (statsEnabled)
 			{
-				resultSetsCommand.Append(cmd);
+				stopWatch = new Stopwatch();
+				stopWatch.Start();
 			}
-		}
 
-		protected void DoExecute()
-		{
-			var resultSetsCommand = Session.Factory.ConnectionProvider.Driver.GetResultSetsCommand(Session);
-			CombineQueries(resultSetsCommand);
-
-			var querySpaces = new HashSet<string>(_queries.SelectMany(t => t.GetQuerySpaces()));
-			if (resultSetsCommand.HasQueries)
+			var querySpaces = new HashSet<string>(
+				_queries.SelectMany(q => q.QueryInformation).SelectMany(qi => qi.QuerySpaces));
+			if (querySpaces.Count > 0)
 			{
+				// The auto-flush must be handled before obtaining the commands, because an auto-flush
+				// may have to invalidate cached data which otherwise would cause a command to be skipped.
 				Session.AutoFlushIfRequired(querySpaces);
 			}
 
-			bool statsEnabled = Session.Factory.Statistics.IsStatisticsEnabled;
-			Stopwatch stopWatch = null;
-			if (statsEnabled)
-			{
-				stopWatch = new Stopwatch();
-				stopWatch.Start();
-			}
+			var resultSetsCommand = Session.Factory.ConnectionProvider.Driver.GetResultSetsCommand(Session);
+			CombineQueries(resultSetsCommand, GetCachedResults());
```

Comment:
I could have put `GetCachedResults()` in `CombineQueries`, but I have decided to surface it here as a counterpart to `PutCacheableResults()` later in this method, and to make it more obvious there is a cache get operation here.